src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -1275,6 +1283,11 @@ ExecuteTransactionEnd(bool commit)
 	NodeConnectionEntry *participant;
 	bool completed = !commit; /* aborts are assumed completed */
 
+	if (xactParticipantHash == NULL)
```

Comment:
> we'll likely revisit the way COPY commits as part of connection management

Arrived at the same thought. I was going to try to handle `COPY` separately, realized it would be cleanest to have a new callback for that, then realized we'd have three such callbacks now, then realized I really wanted to unify them, and finally figured we'd be doing that soon enough with the connection/transaction management code. So I'm punting on that.

Leaving this for now.